### ISSUE TITLE
`extraPaths` starting with `~/` on Linux are considered relative

### ISSUE THREAD

**Issue by debonte** [microsoft/pylance-release#3730]
-   Language Server version: 2022.12.11
-   OS and version: Ubuntu 22.04.1 LTS

## Code Snippet

VS Code settings:
```json
"python.analysis.extraPaths": [
    "~/.ipython/profile_default/startup"
]
```

~/.ipython/profile_default/startup/erikd.py:
```python
def addModulePath(path: str):
    pass
```

## Repro Steps

1. Set extraPaths and create .py file as above.
2. In a .py file within your workspace, type: `from erikd import addModulePath`

## Expected behavior

Import is resolved

## Actual behavior

Import cannot be resolved. `~/.ipython/profile_default/startup` is treated as a relative path. Problem is that `getRootLength` returns 0 for this path.

**Comment by erictraut** [comment]
This is the intended behavior currently. The `~` character is a shell construct, and pyright doesn't contain logic to interpret shell constructs. We could decide to change that, but it's a slippery slope. Would we support other shell constructs like shell variable expansion? For which shells?

We might want to look at other language servers (e.g. typescript) to see what they do in this case.

**Comment by debonte** [comment]
✅ mypy
Mypy's [`mypy_path`](https://mypy.readthedocs.io/en/stable/config_file.html#confval-mypy_path) config setting supports `~` and the docs call that out: "User home directory and environment variables will be expanded."

```text
[mypy]
mypy_path = ~/.ipython/profile_default/startup
```

✅ Pyre
Pyre's [`search_path`](https://pyre-check.org/docs/configuration/) config setting supports `~`.

```json
{
  "site_package_search_strategy": "pep561",
  "source_directories": [
    "."
  ],
  "search_path": [
    "~/.ipython/profile_default/startup"
  ]
}
```

:x: TypeScript
TypeScript's [`compilerOptions/paths`](https://www.typescriptlang.org/tsconfig#paths) does not support `~`.

```json
{
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
       "foo/*": ["/home/erikd/foo/*"], // works
       "foo/*": ["~/foo/*"]            // does not work
    }
  }
}
```

**Comment by rchiodo** [comment]
Related: https://github.com/microsoft/pylance-release/issues/3823

**Comment by heejaechang** [comment]
This issue has been fixed in prerelease version 2023.1.31, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2023131-18-January-2023-PreRelease)
